File: 1002 Find Common Characters.py
class Solution:
    def commonChars(self, A: List[str]) -> List[str]:
        dic = {}
        for k in A[0]:
            dic[k] = 1 if k not in dic else dic[k]+1
        for i in range(1, len(A)):
            for k, v in dic.items():
                dic[k] = min(dic[k], A[i].count(k))
        res = []
        for key, value in dic.items():
            for _ in range(value):
                res.append(key)
        return res

''' Use reduce and collections.Counter
import functools
import collections
class Solution:
    def commonChars(self, A: List[str]) -> List[str]:
        dic = functools.reduce(lambda x, y: collections.Counter(x) & collections.Counter(y), A)
        return list(dic.elements())
'''

File: 1002 Find Common Characters.java
class Solution {
    public List<String> commonChars(String[] A) {
        int[] dict = new int[26];
        List<String> res = new ArrayList<String>();
        for (char ch: A[0].toCharArray()) 
            ++dict[ch - 'a'];
        
        for (int i = 1; i < A.length; ++i) {
            int[] next = new int[26];
            
            for (char ch: A[i].toCharArray()) 
                ++next[ch - 'a'];
            
            for (int j = 0; j < 26; ++j) 
                if (dict[j] > next[j]) 
                    dict[j] = next[j];
        }
        
        for (int i = 0; i < 26; ++i) {
            while (dict[i]-- > 0) res.add("" + (char)('a' + i));
        }
        
        return res;
    }
}